lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,111 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+	logrus "github.com/sirupsen/logrus"
+)
+
+const (
+	// PauseTimeBetweenBenchmarks is the time to pause between each benchmark
+	PauseTimeBetweenBenchmarks = time.Second * 5
+)
+
+// Generator provides a standardized way to get successive benchmarks from a consistent interface
+type Generator interface {
+	// Generate advances the Generator to the next generation.
+	// It returns false when the generator no longer has configurations to run.
+	Generate() bool
+	// GetBenchmark returns the benchmark config for the current generation.
+	// If called after Generate() returns false, this will result in an error.
+	GetBenchmark() (context.Context, Config, error)
+	// Benchmark runs the benchmark of receiver type
+	// return an array of Results that contain information about the generations
+	Benchmark([]string, *client.TeleportClient) ([]*Result, error)
+}
+
+// Run is used to run the benchmarks, it is given a generator, command to run,
+// a host, host login, and proxy. If host login or proxy is an empty string, it will
+// use the default login
+func Run(cnf interface{}, cmd, host, login, proxy string) ([]Result, error) {
+	tc, err := makeTeleportClient(host, login, proxy)
+	if err != nil {
+		return nil, err
+	}
+	logrus.SetLevel(logrus.ErrorLevel)
+	switch c := cnf.(type) {
+	case *Linear:
+		return c.Benchmark(context.Background(), cmd, tc)
+
+	}
+	return nil, errors.New("Invalid generator, not of type Linear")
```

Comment:
Let's not capitalize error strings following our [Go code guidelines](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)